memstore/memstore.go

```diff
@@ -252,3 +258,92 @@ func (m *memStoreImpl) TryEvictBatchColumn(table string, shardID int, batchID in
 	utils.GetLogger().Debugf("Successfully evict batch from memstore: table %s, shardID %d, batchID %d, columnID %d", table, shardID, batchID, columnID)
 	return true, nil
 }
+
+func (m *memStoreImpl) AddTableShard(table string, shardID int, needPeerCopy bool) {
+	m.Lock()
+	defer m.Unlock()
+
+	schema := m.TableSchemas[table]
+	if schema != nil {
+		// table might get deleted at this point
+		return
+	}
+
+	shardMap := m.TableShards[table]
+	if shardMap == nil {
+		shardMap = make(map[int]*TableShard)
+	}
+	if _, exist := shardMap[shardID]; !exist {
+		// create new shard
+		tableShard := NewTableShard(schema, m.metaStore, m.diskStore, m.HostMemManager, shardID, m.redologManagerMaster)
+		if needPeerCopy {
+			tableShard.needPeerCopy = 1
+		}
+		shardMap[shardID] = tableShard
+		utils.AddTableShardReporter(table, shardID)
+	}
+	m.TableShards[table] = shardMap
+}
+
+func (m *memStoreImpl) RemoveTableShard(table string, shardID int) {
+	var shard *TableShard
+	// Detach first.
+	m.Lock()
+	shards := m.TableShards[table]
+	if shards != nil {
+		shard = shards[shardID]
+		delete(shards, shardID)
+		utils.DeleteTableShardReporter(table, shardID)
+	}
+	m.Unlock()
+	// Destruct.
+	if shard != nil {
+		shard.Destruct()
+	}
+}
+
+// preloadAllFactTables preloads recent days data for all columns of all table shards into memory.
+// The number of preloading days is defined at each column level. This call will happen at
+// shard initialization stage.
+func (m *memStoreImpl) preloadAllFactTables() {
```

Comment:
not new code, move preloadAll() within host memory manager.Start() out to the memstore level